pandas/_libs/tslibs/conversion.pyx

```diff
@@ -154,45 +161,48 @@ cpdef inline (int64_t, int) precision_from_unit(str unit):
     """
     cdef:
         int64_t m
+        int64_t multiplier
         int p
         NPY_DATETIMEUNIT reso = abbrev_to_npy_unit(unit)
 
+    if out_reso == NPY_DATETIMEUNIT.NPY_FR_ns:
+        multiplier = 1_000_000_000
+    elif out_reso == NPY_DATETIMEUNIT.NPY_FR_us:
+        multiplier = 1_000_000
+    elif out_reso == NPY_DATETIMEUNIT.NPY_FR_ms:
+        multiplier = 1_000
+    elif out_reso == NPY_DATETIMEUNIT.NPY_FR_s:
+        multiplier = 1
+    else:
+        raise ValueError(f"Invalid out_reso: {out_reso}")
+
     if reso == NPY_DATETIMEUNIT.NPY_FR_Y:
         # each 400 years we have 97 leap years, for an average of 97/400=.2425
         #  extra days each year. We get 31556952 by writing
         #  3600*24*365.2425=31556952
-        m = 1_000_000_000 * 31556952
-        p = 9
+        m = multiplier * 31556952
     elif reso == NPY_DATETIMEUNIT.NPY_FR_M:
         # 2629746 comes from dividing the "Y" case by 12.
-        m = 1_000_000_000 * 2629746
-        p = 9
+        m = multiplier * 2629746
```

Comment:
I'm pretty sure the M and Y cases in precision_from_unit are no longer reached, in which case the entire `m` part of this function i think can be replaced with `get_conversion_factor` (OK to consider out of scope)